explorer/interpreter/interpreter.cpp

```diff
@@ -663,6 +677,35 @@ auto Interpreter::InstantiateWitness(Nonnull<const Witness*> witness)
   return cast<Witness>(value);
 }
 
+auto Interpreter::ConvertStructToClass(
+    Nonnull<const StructValue*> init_struct,
+    Nonnull<const NominalClassType*> class_type, SourceLocation source_loc)
+    -> ErrorOr<Nonnull<NominalClassValue*>> {
+  std::vector<NamedValue> struct_values;
+  std::optional<Nonnull<const NominalClassValue*>> base_instance;
+  // Instantiate the `destination_type` to obtain the runtime
+  // type of the object.
+  CARBON_ASSIGN_OR_RETURN(Nonnull<const Value*> inst_class,
+                          InstantiateType(class_type, source_loc));
```

Comment:
Future work (not something I want to see addressed in this PR):

I think this instantiation step should be done higher up the call stack, by the caller of `Convert` -- generally speaking, when the interpreter pulls a type out of the AST (eg, when looking at the `static_type` of an expression), it should immediately instantiate the type to find the proper interpretation of that type in the current context. We shouldn't be passing around uninstantiated types, because that will lead to a lack of clarity over whose responsibility it is to instantiate and to risks that types will either be instantiated repeatedly or that instantiation will be forgotten. (Doing the instantiation in the `Step` function also prepares us for switching to doing non-recursive type instantiation as an `Action`.)

The status quo seems to be that some callers of `Convert` call `InstantiateType`, and others don't, so we sometimes instantiate the type twice (when converting to a class type), sometimes once, and sometimes zero times, depending on what we're converting to, and in which context.

This is a pre-existing problem, not one introduced in this pull request, but this change does seem to be making it somewhat more pressing, given that the call to `Convert` below, when converting the base class, can recursively land back in this function, meaning that in a depth-N inheritance hierarchy we can instantiate N^2 class types.